lib/serialization/inspect.nit

```diff
@@ -0,0 +1,292 @@
+# This file is part of NIT ( http://www.nitlanguage.org ).
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# Refine `Serializable::inspect` to show more useful information
+module inspect
+
+import serialization_core
+private import caching
+
+# Serialization engine writing the object attributes to strings
+private class InspectSerializer
+	super CachingSerializer
+
+	# Target writing stream
+	var stream: Writer
+
+	redef var current_object = null
+
+	var first_object: nullable Object = null
+
+	redef fun serialize(object)
+	do
+		if object == null then
+			stream.write "null"
+		else
+			if current_object == null then
+				first_object = object
+			end
+
+			var last_object = current_object
+			current_object = object
+			object.accept_inspect_serializer self
+			current_object = last_object
+		end
+	end
+
+	var first_attribute_serialized = false
+
+	redef fun serialize_attribute(name, value)
+	do
+		if first_attribute_serialized then
+			stream.write ", "
+		else
+			stream.write " "
+			first_attribute_serialized = true
+		end
+
+		stream.write name
+		stream.write ":"
+
+		super
+	end
+
+	redef fun serialize_reference(object)
+	do
+
+		if cache.has_object(object) then
+			# Cycle
+			var id = cache.id_for(object)
+			stream.write "<"
+			stream.write object.class_name
+			stream.write "#"
+			stream.write id.to_s
+			stream.write ">"
+		else if object != first_object and (not object isa DirectSerializable) then
+			# Another object, print class and id only
+			var id = cache.new_id_for(object)
+			stream.write "<"
+			stream.write object.class_name
+			stream.write "#"
+			stream.write id.to_s
+			stream.write ">"
+		else
+			# Main object
+			serialize object
+		end
+	end
+end
+
+redef class Serializable
+
+	# Improve the default inspection reading serializable attributes
+	#
+	# Simple immutable data are inspected as they would be written in Nit code:
+	#
+	# ~~~
+	# assert 123.inspect == "123"
+	# assert 1.5.inspect == "1.5"
+	# assert 0xa1u8.inspect == "0xa1u8"
+	# assert 'c'.inspect == "'c'"
+	# assert "asdf\n".inspect == "\"asdf\\n\""
+	# ~~~
+	#
+	# Inspections of mutable object show their dynamic type and an id unique
+	# to each call to `inspect`.
```

Comment:
I'm not sure I like the id unique to each call. The original reason why inspect use the object_id is to be able to correlate the sameness (or the lack of sameness) of objects from various points of inspection. For instance it is helpful to see that the object poped is not the same as the one pushed (even if they have comparable values).

Reply:
I'm not sure either. The main advantage of this "unique id per call" instead of using `object_id` was to make the output reproducible and testable. Maybe I could use `object_id`, but switch to unique id per call when `NIT_TESTING`, what do you think?

Reply:
why not. this seems overkill but works I think.